### ISSUE TITLE
Auto PM breaks requests created from NSURLSession(sessionWithConfiguration:delegate:delegateQueue)

### ISSUE THREAD

**Issue by Kaelzs** [getsentry/sentry-cocoa#1288]
### Environment

How do you use Sentry?
Sentry SaaS (sentry.io)

Which SDK and version?
Sentry 7.2.2

### Steps to Reproduce

[SentryNetworkIssue.zip](https://github.com/getsentry/sentry-cocoa/files/7052431/SentryNetworkIssue.zip)
Using this project to reproduce the issue, just download and hit run. (or comment the SentrySDK.start code in `AppDelegate.swift` and then check the result)

### Expected Result
The request should be block by custom protocol

### Actual Result
The request returned 200.

I've written some [article](https://blog.kaelzs.com/2020/ios-network-debugging-d0abf64b) about network hook in Chinese, here's the [google translated version](https://translate.google.com/translate?sl=zh-CN&tl=en&u=https://blog.kaelzs.com/2020/ios-network-debugging-d0abf64b)

In simple terms, the hook method in sentry SDK (7.2.2) has some issues.

1. It will create a session for every request without any reuse.
2. Each session using the default session configuration, which will override the original configuration
3. Some delegate function in session delegate may not be called (did receive challenge handling)

And I will submit a pull request in a few days to fix it.

**Comment by Kaelzs** [comment]
sounds great! I tried using a demux to make a "fake" session for every different session, but I got stuck by giving the right challenge callback since there's no client callback with a completion block in URLProtocol. As you mentioned, the only reason you using this hook is to add a header in every request, it's better to try hooking URLRequest than breaking the whole request loading system.

**Comment by philipphofmann** [comment]
In a call, @brustolin and I agreed that:

1. we should remove the sizzling of `NSURLSession(sessionWithConfiguration:delegate:delegateQueue)` because we couldn't identify a proper workaround.
2. we add a feature flag for adding the Sentry Trace HTTP header that is turned off by default
3. we add a to-be-defined way of getting the Sentry Trace HTTP header from the SDK so users can add it manually to their custom code. We point that out in the documentation.

**Comment by philipphofmann** [comment]
Alternatives considered: 

1. [Swizzling NSURLRequest](https://github.com/getsentry/sentry-cocoa/pull/1295) didn't work because we need to return an `NSMutableURLRequest` when initializing an `NSURLRequest`, which leads to a crash when creating a request in Swift.
2. Get the customized NSURLSessionConfiguration in `startLoad`, as pointed out by @Kaelzs. Unfortunately, this is not easy and @Kaelzs himself struggled to try to make it work.
3. Swizzling `URLSessionConfiguration.httpAdditionalHeaders`. The problem is that `httpAdditionalHeaders` was not called after sizzling it.
4. Adding the extra header to `URLSessionConfiguration.httpAdditionalHeaders` doesn't work, because the header changes for every transaction.

**Comment by philipphofmann** [comment]
We just figured out that swizzling of `URLSessionConfiguration.httpAdditionalHeaders` works. Most likely, we are going now for this approach instead.

**Comment by philipphofmann** [comment]
Please update the docs after merging this https://github.com/getsentry/sentry-docs/issues/4081.
